File: server/src/main/java/avatar/play/offline/MapOfflineManager.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package avatar.play.offline;

import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author kitakeyos - Hoàng Hữu Dũng
 */
public class MapOfflineManager {

    private static final MapOfflineManager instance = new MapOfflineManager();

    public static MapOfflineManager getInstance() {
        return instance;
    }

    private final List<AbsMapOffline> maps = new ArrayList<>();

    public MapOfflineManager() {
        add(new ObjPremium(3));
        add(new ObjPet(4));
    }

    public void add(AbsMapOffline mapOffline) {
        synchronized (maps) {
            maps.add(mapOffline);
        }
    }

    public void remove(AbsMapOffline mapOffline) {
        synchronized (maps) {
            maps.remove(mapOffline);
        }
    }

    public AbsMapOffline find(int id) {
        synchronized (maps) {
            for (AbsMapOffline mapOffline : maps) {
                if (mapOffline.getId() == id) {
                    return mapOffline;
                }
            }
            return null;
        }
    }
}

File: server-node/src/play/offline/MapOfflineManager.js
/**
 * Port của avatar/play/offline/MapOfflineManager.java
 * @author kitakeyos - Hoàng Hữu Dũng
 */
import { ObjPremium } from './ObjPremium.js';
import { ObjPet } from './ObjPet.js';

export class MapOfflineManager {

  static getInstance() {
    return mapOfflineManager;
  }

  constructor() {
    this.maps = [];
    this.add(new ObjPremium(3));
    this.add(new ObjPet(4));
  }

  add(mapOffline) {
    this.maps.push(mapOffline);
  }

  remove(mapOffline) {
    const idx = this.maps.indexOf(mapOffline);
    if (idx >= 0) {
      this.maps.splice(idx, 1);
    }
  }

  find(id) {
    for (const mapOffline of this.maps) {
      if (mapOffline.getId() === id) {
        return mapOffline;
      }
    }
    return null;
  }
}

export const mapOfflineManager = new MapOfflineManager();
export default mapOfflineManager;
